parl/framework/net.py

```diff
@@ -0,0 +1,236 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import paddle.fluid as fluid
+import parl.layers as layers
+from paddle.fluid.framework import convert_np_dtype_to_dtype_
+
+
+def check_duplicate_spec_names(model):
+    """
+    Check if there are two specs that have the same name.
+    """
+    specs = model.get_input_specs() \
+            + model.get_action_specs() \
+            + model.get_state_specs() \
+            + model.get_reward_specs()
+    names = [name for name, _ in specs]
+    duplicates = set([n for n in names if names.count(n) > 1])
+    assert not duplicates, \
+        "duplicate names with different specs: " + " ".join(duplicates)
+
+
+class Network(object):
+    def copy_to(self, target_net):
+        assert self.__class__.__name__ == target_net.__class__.__name__, \
+            "Only objects of the same Network class can copy between each other"
+        src_layers = self.get_all_layers()
+        target_layers = target_net.get_all_layers()
+        assert set(src_layers.keys()) == set(target_layers.keys()), \
+            "src_layers and target_layers not equal!"
+        for k in src_layers.keys():
+            src_layer = src_layers[k]
+            target_layer = target_layers[k]
+            src_layer.copy_to(target_layer)
+
+    def get_all_layers(self):
+        """
+        For any Network object, this function returns all the class variables
+        that are of the LayerFunc type.
+        Return: a dictionary where the key is the variable str name and the val is
+        the variable.
+        """
+        ## get all the variable names
+        attrs = dir(self)
+        ## only return those whose types are LayerFunc
+        return {attr : getattr(self, attr) for attr in attrs if \
+                isinstance(getattr(self, attr), layers.LayerFunc)}
```

Comment:
should also consider Network objects

Reply:
Currently, I don't consider a hierarchy inside a Network object. When an algorithm is copied, it will always copy the model inside it.